src/libexpr/primops.cc

```diff
@@ -39,6 +39,20 @@ namespace nix {
 InvalidPathError::InvalidPathError(const Path & path) :
     EvalError("path '%s' is not valid", path), path(path) {}
 
+// TODO dedup with copy-paste in installables.cc
+template<typename DP>
+static StorePath getStorePathFrom(const DP & derivedPath)
+{
+    return std::visit(overloaded {
+        [&](const typename DP::Built & bfd) {
+            return getStorePathFrom<SingleDerivedPath>(*bfd.drvPath);
+        },
+        [&](const typename DP::Opaque & bo) {
+            return bo.path;
+        },
+    }, derivedPath.raw());
+}
+
```

Comment:
I was skeptical of making this a method when it just occurred in installables.cc, because it is rarely the right thing and yet awfully tempting to use as the one `DerivedPath -> StorePath` (no `Store`, no extra IO) function. But now that we are using it twice, I guess it should be...